client/db.go

```diff
@@ -0,0 +1,461 @@
+package client
+
+import (
+	"encoding"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/testutils"
+	gogoproto "github.com/gogo/protobuf/proto"
+)
+
+// Timestamp ...
+type Timestamp struct {
+	WallTime int64
+	Logical  int32
+}
+
+// KeyValue ...
+type KeyValue struct {
+	timestamp Timestamp
+	key       []byte
+	value     interface{}
+}
+
+// Key ...
+func (kv *KeyValue) Key(k interface{}) error {
+	switch t := k.(type) {
+	case encoding.BinaryUnmarshaler:
+		return t.UnmarshalBinary(kv.key)
+	case *[]byte:
+		*t = kv.key
+	case *string:
+		*t = string(kv.key)
+	}
+	return fmt.Errorf("unable to unmarshal value: %T", k)
+}
+
+// KeyBytes ...
+func (kv *KeyValue) KeyBytes() []byte {
+	return kv.key
+}
+
+// KeyString ...
+func (kv *KeyValue) KeyString() string {
+	return string(kv.key)
+}
+
+// Value ...
+func (kv *KeyValue) Value(v interface{}) error {
+	switch val := kv.value.(type) {
+	case []byte:
+		switch t := v.(type) {
+		case encoding.BinaryUnmarshaler:
+			return t.UnmarshalBinary(val)
+		case gogoproto.Message:
+			return gogoproto.Unmarshal(val, t)
+		case *[]byte:
+			*t = val
+		case *string:
+			*t = string(val)
+		}
+	case *int64:
+		switch t := v.(type) {
+		case *int64:
+			*t = *val
+		case *uint64:
+			*t = uint64(*val)
+		case *int32:
+			*t = int32(*val)
+		case *uint32:
+			*t = uint32(*val)
+		case *int:
+			*t = int(*val)
+		case *uint:
+			*t = uint(*val)
+		}
+	}
+	return fmt.Errorf("unable to unmarshal value: %T -> %T", kv.value, v)
+}
+
+func (kv *KeyValue) setValue(v *proto.Value) {
+	if v == nil {
+		return
+	}
+	if v.Bytes != nil {
+		kv.value = v.Bytes
+	} else if v.Integer != nil {
+		kv.value = v.Integer
+	}
+	if ts := v.Timestamp; ts != nil {
+		kv.timestamp.WallTime = ts.WallTime
+		kv.timestamp.Logical = ts.Logical
+	}
+}
+
+// ValueBytes ...
+func (kv *KeyValue) ValueBytes() []byte {
+	return kv.value.([]byte)
+}
+
+// ValueString ...
+func (kv *KeyValue) ValueString() string {
+	return string(kv.value.([]byte))
```

Comment:
Oh, good point. That would help if the result from `ValueStringer()` is passed as an `interface{}`, but if someone wants a `string` the cast will be simpler. Anyways, I've added: folks can see what it looks like.